pytket/extensions/cutensornet/tnstate/ttn.py

```diff
@@ -0,0 +1,845 @@
+            of the given root path in the TTN.
+        qubit_position (dict[pytket.circuit.Qubit, tuple[RootPath, int]]): A dictionary
+            mapping circuit qubits to their address in the TTN.
+        fidelity (float): A lower bound of the fidelity, obtained by multiplying
+            the fidelities after each contraction. The fidelity of a contraction
+            corresponds to ``|<psi|phi>|^2`` where ``|psi>`` and ``|phi>`` are the
+            states before and after truncation (assuming both are normalised).
+    """
+
+    def __init__(
+        self,
+        libhandle: CuTensorNetHandle,
+        qubit_partition: dict[int, list[Qubit]],
+        config: Config,
+    ):
+        """Initialise a TTN on the computational state ``|0>``.
+
+        Note:
+            A ``libhandle`` should be created via a ``with CuTensorNet() as libhandle:``
+            statement. The device where the TTN is stored will match the one specified
+            by the library handle.
+
+            The current implementation requires the keys of ``qubit_partition`` to be
+            integers from ``0`` to ``2^l - 1`` for some ``l``.
+
+        Args:
+            libhandle: The cuTensorNet library handle that will be used to carry out
+                tensor operations on the TTN.
+            qubit_partition: A partition of the qubits in the circuit into disjoint
+                groups, describing the hierarchical structure of the TTN. As a rule of
+                thumb, the cost of a gate between qubits on ``qubit_partition[i]`` and
+                ``qubit_partition[j]`` scales exponentially on ``i XOR j``.
+            config: The object describing the configuration for simulation.
+
+        Raises:
+            ValueError: If the keys of ``qubit_partition`` do not range from ``0`` to
+                ``2^l - 1`` for some ``l``.
+            ValueError: If a ``Qubit`` is repeated in ``qubit_partition``.
+            ValueError: If there is only one entry in ``qubit_partition``.
+            NotImplementedError: If the value of ``truncation_fidelity`` in ``config``
+                is smaller than one.
+        """
+        self._lib = libhandle
+        self._cfg = config
+        self._logger = set_logger("TTN", level=config.loglevel)
+        self.fidelity = 1.0
+        self.nodes: dict[RootPath, TreeNode] = dict()
+        self.qubit_position: dict[Qubit, tuple[RootPath, int]] = dict()
+
+        if self._cfg.truncation_fidelity < 1:
+            raise NotImplementedError(
+                "Truncation fidelity mode not currently implemented on TTN."
+            )
+
+        n_groups = len(qubit_partition)
+        if n_groups == 0:  # There's no initialisation to be done
+            return None
+        if n_groups == 1:
+            raise ValueError(
+                "Only one entry to qubit_partition provided."
+                "Introduce a finer partition of qubits."
+            )
+
+        n_levels = math.floor(math.log2(n_groups))

```

Comment:
Hmm I think I'd need to understand this... Why if you have 4 groups, you'd have 2 levels? I'd thought in a binary tree you should have three?